crates/extensions/tedge_mqtt_ext/src/test_helpers.rs

```diff
@@ -48,14 +33,88 @@ pub async fn assert_received_includes_json<I, S>(
 {
     for expected_msg in expected.into_iter() {
         let message = messages.recv().await.expect("MQTT channel closed");
-        assert_eq!(message.topic, Topic::new_unchecked(expected_msg.0.as_ref()));
-        let payload = serde_json::from_str::<serde_json::Value>(
-            message.payload_str().expect("non UTF-8 payload"),
-        )
-        .expect("non JSON payload");
-        assert_json_include!(
-            actual: payload,
-            expected: expected_msg.1
-        );
+        assert_message_includes_json(&message, expected_msg);
+    }
+}
+
+pub fn assert_messages_contains_str<I, S>(messages: &mut Vec<MqttMessage>, expected: I)
```

Comment:
Instead of  restricting the inputs to `&mut Vec<MqttMessage>`, I'd like to open up this API to accept any iterable that supports in-place removal of items from the source collection. It is taking an `&mut` input to remove all matched items from the input and leave the remaining items in the source. It is done this way to support the re-use of the source for further assertions, like the following sequence:

```rust
let messages = // Converted Vec<Message>
assert_messages_contains_str(
            &mut messages, ...)  // Initial messages
assert_messages_includes_json(
            &mut messages, ...)  // Remaining messages
```

So, all suggestions welcome.

Reply:
I think what you're looking for here is just to take an iterator directly and call `.next()` instead of `.remove(0)`. That would also allow you to combine removing with checking whether the iterator is empty, e.g. the body of this function could become:

```rust
    for expected_msg in expected {
        if let Some(message) = messages.next() {
              assert_message_contains_str(&message, expected_msg);
        } else {
             panic!(
                  "Input messages exhausted while expecting message on topic: {:?} with payload: {:?}",
                  expected_msg.0.as_ref(), expected_msg.1
              )
        }
    }
```

Personally, I find a function called `assert_...` that mutates its input to be really weird. Additionally when I read `assert_messages_contains_str`, it sounds like it checks for any message in the list that matches a single expectation, rather than comparing the first N elements of the list with a list of N expectations. I wonder whether asserting against an array is all that useful. I don't know in detail what we are currently testing that could make use of this, but it feels like if we're asserting that more than a small handful of messages exist in a single test, that's probably indicative that there's a larger problem with the test being too complicated (which may itself indicate that the code being tested is hard to test).

I find the following considerably simpler and clearer as far as test code is concerned (and it means far less test helper code). Obviously if we don't have as many messages as expected, this won't give as clear an error message in the test, but it will still tell us which assertion failed.

```rust
let messages = messages.into_iter();

assert_message_contains_str(
    messages.next().unwrap(),
    (
        "c8y/s/us",
        "101,test-device:device:child1,child1,thin-edge.io-child",
    ),
);

assert_message_includes_json(
    messages.next().unwrap(),
    (
        "c8y/measurement/measurements/create",
        json!({
            "externalSource":{
                "externalId":"test-device:device:child1",
                "type":"c8y_Serial"
            },
            "temp":{
                "temp":{
                    "value":1.0
                }
            },
            "time":"2021-11-16T17:45:40.571760714+01:00",
            "type":"ThinEdgeMeasurement"
        }),
    ),
);
```

Reply:
I agree with @jarhodes314 that the name `assert_messages_contains_str` is misleading.

I also wonder why [`tedge_actores::tedge_helpers::assert_received_matching`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_actors/src/test_helpers.rs#L239) cannot be used instead of a specific `assert_messages_contains_str`

Reply:
@jarhodes314 

> I think what you're looking for here is just to take an iterator directly and call .next() instead of .remove(0)

Yeah, this would have been possible if **all** the expected messages were provided together as the input and then I could have iterated over the entire collection. But, the limitation here is that, I couldn't provide all the inputs together in a single `expected` array as some of them were just plain strings and others were JSON values. This is why we're doing partial matches of the whole collection in different iterations with a combination of `assert_messages_contains_str` and `assert_messages_contains_json`. One way to avoid this is to have an enum as follows to represent the expected messages and accept an array of these for the assertions:

```rust
pub enum MqttPayload {
    StringMessage(String),
    JsonMessage(serde_json::Value),
    Skip(usize),
}

pub fn assert_messages<I, S>(messages: &mut Vec<MqttMessage>, expected: I)
where
    I: for<'a> IntoIterator<Item = (&'static str, MqttPayload)>,
{ ... }
```

> Personally, I find a function called assert_... that mutates its input to be really weird

Yeah, I agree, especially for an input like a `Vec`. I was just trying to replicate the same API style that we had for the existing `assert_received_xxx` functions. That probably wasn't the best idea.

> Additionally when I read assert_messages_contains_str, it sounds like it checks for any message in the list that matches a single expectation, rather than comparing the first N elements of the list with a list of N expectations

Agree with that too. I'll try to come up with a better name. Naming suggestions welcome.

> I wonder whether asserting against an array is all that useful. I don't know in detail what we are currently testing that could make use of this, but it feels like if we're asserting that more than a small handful of messages exist in a single test, that's probably indicative that there's a larger problem with the test being too complicated (which may itself indicate that the code being tested is hard to test).

Yeah, we've got a few tests that assert upto 6 messages(all the init messages generated by the mapper). Even in most normal tests, we typically assert upto 4-5 messages(usually a device registration TEdge JSON message along with its SmartREST mapping followed by the actual converted message(s)).

> I find the following considerably simpler and clearer as far as test code is concerned

Looking at that code, yes, I'd also agree with you. But, with the above proposed enum combined with the inbuilt `Into` impls that you suggested for those enum variants, I think I can provide a better API that can accept the entire list of expected inputs.

@didier-wenzek 

> I also wonder why [tedge_actores::tedge_helpers::assert_received_matching](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_actors/src/test_helpers.rs#L239) cannot be used instead of a specific assert_messages_contains_str

If we're asserting messages in isolation, as shown in his sample, yes this definitely is an option. The goal of the helper function is to even eliminate the need for the user to provide the matching logic, just as a list out desired output messages.